pkg/trie/triedb/node.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package triedb
+
+import (
+	"bytes"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/pkg/trie/db"
+	"github.com/ChainSafe/gossamer/pkg/trie/triedb/codec"
+)
+
+type StorageHandle struct{ int }
+
+func (sh StorageHandle) toNodeHandle() NodeHandle {
+	return InMemory{idx: sh}
+}
+
+type NodeHandle interface {
+	isNodeHandle()
+}
+
+type (
+	InMemory struct {
+		idx StorageHandle
+	}
+	Hash struct {
+		hash common.Hash
+	}
+)
+
+func (InMemory) isNodeHandle() {}
+func (Hash) isNodeHandle()     {}
+
+func newFromEncodedMerkleValue(
+	parentHash common.Hash,
+	encodedNodeHandle codec.MerkleValue,
+	storage NodeStorage,
+) (NodeHandle, error) {
+	switch encoded := encodedNodeHandle.(type) {
+	case codec.HashedNode:
+		return Hash{hash: common.NewHash(encoded.Data)}, nil
+	case codec.InlineNode:
+		child, err := newNodeFromEncoded(parentHash, encoded.Data, storage)
+		if err != nil {
+			return nil, err
+		}
+		return InMemory{storage.alloc(New{child})}, nil
+	default:
+		panic("unreachable")
+	}
+}
+
+type Value interface {
+	isValue()
+	getHash() common.Hash
+}
+
+type (
+	Inline struct {
+		Data []byte
+	}
+
+	ValueRef struct {
+		hash common.Hash
+	}
+
+	NewValueRef struct {
+		hash *common.Hash
+		Data []byte
+	}
+)
+
+func (Inline) isValue()                  {}
+func (Inline) getHash() common.Hash      { return common.EmptyHash }
```

Comment:
I guess this is similar to the types in Actions, if we have `getHash()` on all of these do we still need `isValue()`? Is it just to more clearly demonstrate?

Reply:
I can remove `isValue()` 😄